Implement new trigger filters experimental feature (SubscriptionsAPI Filters)

devguyio opened the pull request:
Implementation of the new filters experimental feature

Based on #1921 , #1977 & #1978

Fixes #1763 

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

- Implement SubscriptionsAPI Filters / aka new trigger filters experimental feature.

## TODO
- [x] Contract tests


<!--
If this change has user-visible impact, follow the instructions below.
Examples include:

- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic

Otherwise delete the rest of this template.
-->

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release-note is needed.
-->

```release-note
- 🎁 Add new `new-trigger-filters` experimental feature. When enabled, Triggers support a new `filters` field that conforms to the filters API field defined in the [`CloudEvents Subscriptions API`](https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#324-filters). It allows you to specify a set of powerful filter expressions, where each expression evaluates to either true or false for each event.
```

devguyio commented:
@matzew @aliok @pierDipi you can follow the commits flow for reviewing.

aliok commented:
/rebase

matzew commented:
> /rebase

does that really work ? :joy:

matzew commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/AllFilter.java:
```diff
@@ -13,10 +13,11 @@
  * limitations under the License.
  */
-package dev.knative.eventing.kafka.broker.dispatcher.impl.filter;
+package dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi;
```
:heart_eyes:

matzew commented on data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerImpl.java:
```diff
@@ -22,6 +22,8 @@
 import dev.knative.eventing.kafka.broker.core.utils.CollectionsUtils;
 import io.vertx.core.CompositeFuture;
 import io.vertx.core.Future;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
nit: but looks like some import setting seems to be different

devguyio replied:
Fixed

matzew commented:
I think, as said on the other PR this looks good

I had only a nit, do to import/format changes.

aliok commented on control-plane/pkg/reconciler/trigger/trigger.go:
```diff
@@ -63,6 +67,7 @@ func (r *Reconciler) ReconcileKind(ctx context.Context, trigger *eventing.Trigge
 }
 
 func (r *Reconciler) reconcileKind(ctx context.Context, trigger *eventing.Trigger) reconciler.Event {
+	ctx = r.WithCtx(ctx)
```
I've checked knative/eventing code and I see the feature store is used that way. But I found it strange. Why don't the feature store is set into the ctx instead of creating a wrapper function? Is it for testing?

aliok replied:
I am not "asking" you to change it, but I don't get why it is done that way in knative/eventing repo as well.

devguyio replied:
@aliok I updated the approach and now there's a callback that gets triggered by the ConfigWatcher which just updates the cached feature flags in the reconciler. I removed the WithCtx.

aliok commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/main/ConsumerVerticleFactoryImpl.java:
```diff
@@ -148,72 +155,99 @@ public AbstractVerticle get(final DataPlaneContract.Resource resource, final Dat
-                .setConfig(consumerConfigs)
-                // Make sure the policy is propagate for the manually instantiated consumer tracer
-                .setTracingPolicy(TracingPolicy.PROPAGATE)
+          KafkaClientsAuth.attachCredentials(consumerConfigs, credentials);
+          KafkaClientsAuth.attachCredentials(producerConfigs, credentials);
+
+          final KafkaConsumer<Object, CloudEvent> consumer = createConsumer(vertx, consumerConfigs);
+          AutoCloseable metricsCloser = Metrics.register(consumer.unwrap());
```
Mind checking your IDE settings for the project to use the `.editorconfig` file in the repo?

devguyio replied:
Done. I had to manually revert though, that change is not related to the .editorconfig, it's some setting in Intellij

aliok commented on .github/workflows/kind-e2e.yaml:
```diff
@@ -63,6 +64,12 @@ jobs:
             channel-auth-scenario: "SASL_SSL"
           - test-suite: ./test/e2e_new
             channel-auth-scenario: "SASL_PLAIN"
+          - test-suite: ./test/experimental
+            channel-auth-scenario: "SSL"
+          - test-suite: ./test/experimental
+            channel-auth-scenario: "SASL_SSL"
+          - test-suite: ./test/experimental
+            channel-auth-scenario: "SASL_PLAIN"
```
These are only for testing KafkaChannel with auth. The reasoning is:
- We cannot define individual KafkaChannels to use auth as it is a global thing
- Thus, we re-run the entire KafkaChannel reconciler test suite in a separate job with auth enabled.

You can remove these lines here.

matzew replied:
agreed, and yes - we have upstream... also several jobs

devguyio replied:
@aliok @matzew I'm not sure I follow. What I'm doing here is excluding running the experimental features suite as part of the KafkaChannel e2e jobs. This insures that for the SSL, SASL_SSL & SASL_PLAIN scenarios (i.e. the 3 KafkaChannel scenarios) only the test-suite `./test/e2e_channel` will be picked up, and `./test/experimental` will be excluded.

matzew replied:
Context FTW :smile: 

looking _at_ the file directly, I do see `exclude` a couple of lines above.

aliok replied:
> looking at the file directly, I do see exclude a couple of lines above.

My bad. I am sorry 🤦

aliok commented on test/experimental/main_test.go:
```diff
@@ -0,0 +1,65 @@
+//go:build e2e
```
I like having the tests for experimental features in a separate directory!

aliok commented on test/experimental/features_config/install.go:
```diff
@@ -0,0 +1,32 @@
+	"knative.dev/reconciler-test/pkg/manifest"
+)
+
+const ConfigMapName = "config-features"
```
Not used

devguyio replied:
Fixed

aliok commented on test/experimental/features_config/install.go:
```diff
@@ -0,0 +1,32 @@
+
+const ConfigMapName = "config-features"
+
+func Install(ctx context.Context, t feature.T) {
```
Can you add a test for this like in here? : https://github.com/knative-sandbox/eventing-kafka/blob/b42f8dff9753f65aa41a3459a75f898d8366578c/test/rekt/resources/kafkachannel/kafkachannel_test.go

aliok replied:
Oh, no need. There are nothing configurable in the YAML anyway. Never mind.

matzew commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,196 @@
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Cesql{
+			Cesql: &CESQL{
+				Expression: f.SQL,
```
While `f.SQL` currently works (we have blockers on the dependency updates).

Soon this will be `CESQL` - but that's more a FYI (especially with a funky state for dependency updates, upstream :see_no_evil: )

https://github.com/knative/eventing/pull/6148/files#diff-9897c649cc6cb820cddfd787a041f960074f91d38b87bd24216e110e76718c1eR175

devguyio replied:
Yep

matzew commented:
LGTM, as already stated on the other PR.

pierDipi commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,196 @@
+	for k, v := range f.Prefix {
+		p.Attribute = k
+		p.Prefix = v
+		// We only expect a single element, but let's break anyway
+		break
+	}
+
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Prefix{
+			Prefix: p,
+		},
+	}
+}
+
+// NewSuffixFilter converts the SubscriptionsAPIFilter into the suffix dialect of
+// the DialectedFilter as defined in CloudEvents Subscriptions API.
+//
+// Suffix contains exactly one attribute where the key is the name of the CloudEvent
+// attribute which value must end with the value specified.
+//
+// See "CNCF CloudEvents Subscriptions API" > "3.2.4.1 Filters Dialects"
+// https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#suffix-filter-dialect
+func NewSuffixFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+
+	p := &Suffix{}
+	// Eventing Trigger API is conforming to the Subscriptions API, so this map has only a single
+	// key-value pair
+	for k, v := range f.Suffix {
+		p.Attribute = k
+		p.Suffix = v
+		// We only expect a single element, but let's break anyway
+		break
+	}
+
+	return &DialectedFilter{
+		Filter: &DialectedFilter_Suffix{
+			Suffix: p,
+		},
+	}
+}
```
As the other comment on the PR, there shouldn't be a difference between these 3 filters.

devguyio replied:
Fixed

pierDipi commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,196 @@
+//
+// E.g a SubscriptionsAPIFilter with the All field set will be converted to a DialectedFilter
+// which Filter is of type DialectedFilter_All.
+func FromSubscriptionFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+	switch {
+	case len(f.All) > 0:
+		return NewAllFilter(f.All)
+	case len(f.Any) > 0:
+		return NewAnyFilter(f.Any)
+	case f.Not != nil:
+		return NewNotFilter(*f.Not)
+	case f.Exact != nil:
+		return NewExactFilter(f)
+	case f.Prefix != nil:
+		return NewPrefixFilter(f)
+	case f.Suffix != nil:
+		return NewSuffixFilter(f)
+	case f.SQL != "":
+		return NewCESQLFilter(f)
+	}
+	return &DialectedFilter{}
+}
```
Can we decouple the one filter constraint from the contract?
I remember making this comment in another PR :thinking:

pierDipi replied:
In addition, the CE subscription spec doesn't have this constraint, so I would really avoid designing a contract in this way

devguyio replied:
> Can we decouple the one filter constraint from the contract? I remember making this comment in another PR thinking

Yes I'm working on it. 

> In addition, the CE subscription spec doesn't have this constraint
Yeah while it's not in the spec itself, it's in the reference openapi file and I remember that it was brought up in one of the PRs introducing the spec. But I agree that we shouldn't have it tightly coupled here.
https://github.com/cloudevents/spec/blob/main/subscriptions/subscriptions-openapi.yaml#L175-L183

devguyio replied:
Fixed

pierDipi commented on data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/SqlFilterTest.java:
```diff
@@ -13,17 +13,18 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
 import java.net.URI;
 import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.util.stream.Stream;
-import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.Arguments;
-import org.junit.jupiter.params.provider.MethodSource;
```
Can we revert these changes?

devguyio replied:
Done

pierDipi commented on data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/SuffixFilterTest.java:
```diff
@@ -13,17 +13,18 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
 import java.net.URI;
 import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.util.stream.Stream;
-import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.Arguments;
-import org.junit.jupiter.params.provider.MethodSource;
```
Can we revert these changes?

devguyio replied:
Done

pierDipi commented on data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/PrefixFilterTest.java:
```diff
@@ -13,17 +13,18 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
 import java.net.URI;
 import java.time.OffsetDateTime;
 import java.time.ZoneOffset;
 import java.util.stream.Stream;
-import org.junit.jupiter.params.ParameterizedTest;
-import org.junit.jupiter.params.provider.Arguments;
-import org.junit.jupiter.params.provider.MethodSource;
```
Can we revert these changes?

devguyio replied:
Done

pierDipi commented on data-plane/dispatcher/src/test/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/ExactFilterTest.java:
```diff
@@ -13,19 +13,20 @@
-import java.time.OffsetDateTime;
-import java.time.ZoneOffset;
-import java.util.stream.Stream;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
 import org.junit.jupiter.params.provider.Arguments;
 import org.junit.jupiter.params.provider.MethodSource;
 
+import java.net.URI;
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.util.stream.Stream;
+
```
Can we revert these changes?

devguyio replied:
Done

pierDipi commented on data-plane/core/src/main/java/dev/knative/eventing/kafka/broker/core/reconciler/impl/ResourcesReconcilerImpl.java:
```diff
@@ -250,7 +249,8 @@ private boolean egressEquals(DataPlaneContract.Egress e1, DataPlaneContract.Egre
       && Objects.equals(e1.getFilter(), e2.getFilter())
-      && Objects.equals(e1.getKeyType(), e2.getKeyType());
+      && Objects.equals(e1.getKeyType(), e2.getKeyType())
+      && Objects.equals(e1.getDialectedFilterList(), e2.getDialectedFilterList());
```
Can we add a test for this change detection?

devguyio replied:
Done

matzew commented:
Is the PR still a WIP ?

devguyio commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,196 @@
+// https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#any-filter-dialect
+func NewAnyFilter(filters []v1.SubscriptionsAPIFilter) *DialectedFilter {
+	any := Any{
+		Filters: []*DialectedFilter{},
```
Fixed

devguyio commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,196 @@
+// https://github.com/cloudevents/spec/blob/main/subscriptions/spec.md#all-filter-dialect
+func NewAllFilter(filters []v1.SubscriptionsAPIFilter) *DialectedFilter {
+	all := All{
+		Filters: []*DialectedFilter{},
```
Fixed

devguyio commented:
> Is the PR still a WIP ?

@matzew @aliok @pierDipi the PR is now ready. All comments were addressed. I rewrote the commits to make them easy to review so take your time going through them commit by commit.

devguyio commented:
The logging formatting I introduced makes it easy to debug failed filters. Here's a sample output with log level set to DEBUG 
```
stern -n knative-eventing kafka-broker-dispatcher -i "impl.filter" --json-pp
```

![2022-02-25_13-38](https://user-images.githubusercontent.com/1930204/155716653-bf4c593c-942b-4fa0-9d48-f2d50c22206b.png)

devguyio commented:
/retest-required

devguyio commented:
/retest-required

Not sure if this is a flake
```
=== RUN   TestAsyncProber/single_pod
    logger.go:130: 2022-02-25T12:53:32.644Z	DEBUG	prober/prober.go:63	Sending probe request	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1"}
    logger.go:130: 2022-02-25T12:53:32.655Z	DEBUG	prober/prober.go:63	Sending probe request	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1"}
    logger.go:130: 2022-02-25T12:53:32.733Z	DEBUG	prober/prober.go:63	Sending probe request	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1"}
    logger.go:130: 2022-02-25T12:53:32.735Z	DEBUG	prober/prober.go:63	Sending probe request	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1"}
    async_prober_test.go:144: 
        	Error Trace:	async_prober_test.go:144
        	Error:      	Condition never satisfied
        	Test:       	TestAsyncProber/single_pod
    logger.go:130: 2022-02-25T12:53:34.441Z	DEBUG	prober/prober.go:63	Sending probe request	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1"}
    logger.go:130: 2022-02-25T12:53:34.442Z	ERROR	prober/prober.go:81	Failed probe	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "error": "Get \"http://127.0.0.1:43661/b1/b1\": dial tcp 127.0.0.1:43661: connect: connection refused"}
    logger.go:130: 2022-02-25T12:53:34.443Z	DEBUG	prober/async_prober.go:128	Probe status	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "status": "Unknown"}
    logger.go:130: 2022-02-25T12:53:34.443Z	ERROR	prober/prober.go:81	Failed probe	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "error": "Get \"http://127.0.0.1:43661/b1/b1\": dial tcp 127.0.0.1:43661: connect: connection reset by peer"}
    logger.go:130: 2022-02-25T12:53:34.443Z	DEBUG	prober/async_prober.go:128	Probe status	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "status": "Unknown"}
    logger.go:130: 2022-02-25T12:53:34.443Z	ERROR	prober/prober.go:81	Failed probe	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "error": "Get \"http://127.0.0.1:43661/b1/b1\": dial tcp 127.0.0.1:43661: connect: connection refused"}
    logger.go:130: 2022-02-25T12:53:34.443Z	DEBUG	prober/async_prober.go:128	Probe status	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "status": "Unknown"}
    logger.go:130: 2022-02-25T12:53:34.444Z	ERROR	prober/prober.go:81	Failed probe	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "error": "Get \"http://127.0.0.1:43661/b1/b1\": dial tcp 127.0.0.1:43661: connect: connection refused"}
    logger.go:130: 2022-02-25T12:53:34.445Z	DEBUG	prober/async_prober.go:128	Probe status	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "status": "Unknown"}
    logger.go:130: 2022-02-25T12:53:34.444Z	ERROR	prober/prober.go:81	Failed probe	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "error": "Get \"http://127.0.0.1:43661/b1/b1\": dial tcp 127.0.0.1:43661: connect: connection refused"}
    logger.go:130: 2022-02-25T12:53:34.445Z	DEBUG	prober/async_prober.go:128	Probe status	{"scope": "prober", "pod.metadata.name": "p1", "address": "http://127.0.0.1:43661/b1/b1", "status": "Unknown"}
    --- FAIL: TestAsyncProber/single_pod (2.81s)
FAIL control-plane/pkg/prober.TestAsyncProber/single_pod (2.81s)
PASS control-plane/pkg/prober.TestAsyncProber/single_pod_-_404 (1.10s)
=== RUN   TestAsyncProber
=== PAUSE TestAsyncProber
=== CONT  TestAsyncProber
--- FAIL: TestAsyncProber (5.01s)
FAIL control-plane/pkg/prober.TestAsyncProber (5.01s)
FAIL control-plane/pkg/prober
```

devguyio commented:
/retest-required

Infra failure, tests actually passed 

```
2022/02/25 13:13:40 Getting artifacts dir from prow
2022/02/25 13:14:26 Failed to run tests with kubetest2: error getting the cluster version: Unable to connect to the server: context deadline exceeded (Client.Timeout exceeded while awaiting headers)
```

pierDipi commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/CeSqlFilter.java:
```diff
@@ -22,23 +22,33 @@
   @Override
   public boolean test(CloudEvent cloudEvent) {
     try {
+      logger.debug("{}: Testing event against CESQL expression. Expression {} - Event {}", this.getClass().getSimpleName(),
+        this.expression, cloudEvent);
       Object value = this.expression.tryEvaluate(this.runtime, cloudEvent);
+      logger.debug("{}: CESQL evaluation succeeded. Expression {} - Event {} - Result {}", this.getClass().getSimpleName(),
+        expression, cloudEvent, value);
       return (Boolean) this.runtime.cast(value, Type.BOOLEAN);
     } catch (EvaluationException evaluationException) {
+      logger.error("{}: Exception while evaluating CESQL expression. Test failed. Expression {} - Event {} - Exception {}",
+        this.expression, this.getClass().getSimpleName(), cloudEvent, evaluationException);
       return false;
```
Same here:
> the logger itself contains the class, there is no need for this.getClass().getSimpleName()

devguyio replied:
Fixed

pierDipi commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/NotFilter.java:
```diff
@@ -0,0 +1,41 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class NotFilter implements Filter {
```
This class doesn't have unit tests, can we add them?

devguyio replied:
Fixed

pierDipi commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/NotFilter.java:
```diff
@@ -0,0 +1,41 @@
+
+  @Override
+  public boolean test(CloudEvent cloudEvent) {
+    logger.debug("{}: Testing NOT filter. Event {}", cloudEvent);
```
Same here:
> Since we're logging passing/failing filter, I think this line can be removed since this context is already logged by other classes (we log something similar before the filters run)

devguyio replied:
Replied above

pierDipi commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/main/ConsumerVerticleFactoryImpl.java:
```diff
@@ -148,72 +154,100 @@ public AbstractVerticle get(final DataPlaneContract.Resource resource, final Dat
+    // Dialected filters should override the attributes filter
+    if (egress.getDialectedFilterCount() > 0) {
+      logger.debug("Egress contains dialected-filters. Dialected-filters will be used for event filtering. Egress {}", egress.getReference());
+      return getFilter(egress.getDialectedFilterList());
+    } else if (egress.hasFilter()) {
```
`else if` after `return`
```suggestion
      return getFilter(egress.getDialectedFilterList());
    }
    if (egress.hasFilter()) {
```

devguyio replied:
I don't see why is this better than the other? care to elaborate? is it a personal preference?

pierDipi replied:
The else is useless if the prev branch has a return and this is the pattern used everywhere.

devguyio replied:
> The else is useless if the prev branch has a return and this is the pattern used everywhere in this codebase.

If it's the pattern used everywhere then it makes sense to be consistent. Fwiw I personally prefer the if/elseif/elseif cause it gives me a better higher level mental model about the codepath without needing to dig into each branch, e.g. if the IDE is folding code blocks.

pierDipi replied:
Ok, fine for me, I see we use both :man_shrugging:

pierDipi commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/main/ConsumerVerticleFactoryImpl.java:
```diff
@@ -148,72 +154,100 @@ public AbstractVerticle get(final DataPlaneContract.Resource resource, final Dat
+    return Filter.noop();
+  }
+
+  private Filter getFilter(DataPlaneContract.DialectedFilter filter) {
+    return switch (filter.getFilterCase()) {
+      case EXACT -> new AttributesFilter(filter.getExact().getAttributesMap());
+      case PREFIX -> new PrefixFilter(filter.getPrefix().getAttributesMap());
+      case SUFFIX -> new SuffixFilter(filter.getSuffix().getAttributesMap());
+      case NOT -> new NotFilter(getFilter(filter.getNot().getFilter()));
+      case ANY -> new AnyFilter(filter.getAny().getFiltersList().stream().
+        map(this::getFilter).collect(Collectors.toSet()));
+      case ALL -> new AllFilter(filter.getAll().getFiltersList().stream().
+        map(this::getFilter).collect(Collectors.toSet()));
+      case CESQL -> new CeSqlFilter(filter.getCesql().getExpression());
+      default -> Filter.noop();
+    };
+  }
+
+  private Filter getFilter(List<DataPlaneContract.DialectedFilter> filters) {
+    return new AllFilter(filters.stream().map(this::getFilter).collect(Collectors.toSet()));
+  }
```
Can these methods be static?

devguyio replied:
Done

pierDipi commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,198 @@
+	}
+}
+
+// FromSubscriptionFilter converts a SubscriptionsAPIFilter to the corresponding
+// DialectedFilter based on the property set in the SubscriptionsAPIFilter. If
+// the SubscriptionsAPIFilter has multiple dialects set, they will be wrapped inside
+// a Filter of the type DialectedFilter_All
+//
+// E.g a SubscriptionsAPIFilter with the Exact field set will be converted to a DialectedFilter
+// which Filter is of type DialectedFilter_Exact.
+func FromSubscriptionFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
```
This is the only function that needs to be public, can we reduce the scope for others `New*Filter` functions?

devguyio replied:
Fixed

pierDipi reviewed (commented):
I left some comments. :+1:

pierDipi commented:
The main concerns on merging this PR, 1 day before the release, are
- the PR is big and touches every component, so periodic jobs might reveal regressions but we don't have time to fix them
- tests are not passing as of today [1]
- I reviewed it again and I have additional comments
- CE SQL implementation isn't stable and deeply tested at all, so even experimenting with this new feature might impact other triggers (for example, the validation in the webhook uses the Go implementation, while our data plane uses the Java one and they might contain subtle differences) in terms of performance and reliability

[1]:
```
--- FAIL: TestNewTriggerFilters (78.38s)
    --- PASS: TestNewTriggerFilters/Prerequisite (3.19s)
        --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/enable_new_trigger_filters_experimental_feature (0.04s)
        --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/install_one_partition_configuration (0.02s)
        --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/install_broker (0.08s)
        --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/broker_is_ready (3.03s)
        --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/broker_is_addressable (0.01s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger_subscriber (3.48s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger (0.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Requirement/Install_matched_event_sender (3.11s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Requirement/Install_unmatched_event_sender (3.15s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (4.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger_subscriber (3.10s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger (0.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Requirement/Install_matched_event_sender (3.17s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Requirement/Install_unmatched_event_sender (3.09s)
    --- FAIL: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger_subscriber (3.15s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger (0.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Requirement/Install_matched_event_sender (3.10s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Requirement/Install_unmatched_event_sender (3.10s)
    --- FAIL: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.01s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger_subscriber (3.14s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger (0.03s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Requirement/Install_matched_event_sender (3.11s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Requirement/Install_unmatched_event_sender (3.14s)
    --- FAIL: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
    --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
```

matzew commented:
/hold

devguyio commented on control-plane/pkg/reconciler/trigger/controller.go:
```diff
@@ -85,6 +87,19 @@ func NewController(ctx context.Context, _ configmap.Watcher, configs *config.Env
+			}
+			logger.Debug("Features ConfigMap updated. Updating cached features.")
+			reconciler.Flags = &flags
+		},
```
Fixed

devguyio commented on control-plane/pkg/reconciler/trigger/controller.go:
```diff
@@ -85,6 +87,19 @@ func NewController(ctx context.Context, _ configmap.Watcher, configs *config.Env
+		func(name string, value interface{}) {
+			flags, ok := value.(feature.Flags)
+			if !ok {
+				logger.Warn("Features ConfigMap updated but we didn't get expected flags. Skipping updating cached features")
```
Done, we get passed the updated configmap name, we don't have access to the actual key, that would need a change in pkg.

devguyio commented on control-plane/pkg/reconciler/trigger/controller.go:
```diff
@@ -85,6 +87,19 @@ func NewController(ctx context.Context, _ configmap.Watcher, configs *config.Env
 	})
 
+	featureStore := feature.NewStore(
+		logging.FromContext(ctx).Named("feature-config-store"),
```
done

devguyio commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,198 @@
+// E.g a SubscriptionsAPIFilter with the Exact field set will be converted to a DialectedFilter
+// which Filter is of type DialectedFilter_Exact.
+func FromSubscriptionFilter(f v1.SubscriptionsAPIFilter) *DialectedFilter {
+	filters := make([]*DialectedFilter, 0, 0)
```
done

devguyio commented on control-plane/pkg/contract/subscriptionsapi.go:
```diff
@@ -0,0 +1,198 @@
+			Any: &any,
+		},
+	}
+}
```
done

devguyio commented:
A maven central glitch, tests actually passed
```
[ERROR] Plugin org.apache.maven.plugins:maven-shade-plugin:3.2.4 or one of its dependencies could not be resolved: Failed to read artifact descriptor for org.apache.maven.plugins:maven-shade-plugin:jar:3.2.4: Could not transfer artifact org.apache.maven.plugins:maven-shade-plugin:pom:3.2.4 from/to central (https://repo.maven.apache.org/maven2): transfer failed for https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-shade-plugin/3.2.4/maven-shade-plugin-3.2.4.pom: Connection reset -> [Help 1]
```

/retest

devguyio commented:
> pull-knative-sandbox-eventing-kafka-broker-integration-tests

Seems like a flaky test that's irrelevant to this PR

```
=== FAIL: test/e2e TestKafkaSinkV1Alpha1AuthSslSaslScram512/0 (9.52s)
    creation.go:500: Creating service account &ServiceAccount{ObjectMeta:{eventing-e2e9  eventing-e2e9    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Secrets:[]ObjectReference{},ImagePullSecrets:[]LocalObjectReference{},AutomountServiceAccountToken:nil,}
    creation.go:528: Creating role &Role{ObjectMeta:{eventing-e2e9      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Rules:[]PolicyRule{PolicyRule{Verbs:[get],APIGroups:[],Resources:[pods],ResourceNames:[],NonResourceURLs:[],},PolicyRule{Verbs:[*],APIGroups:[],Resources:[events],ResourceNames:[],NonResourceURLs:[],},},}
    creation.go:547: Creating role binding &RoleBinding{ObjectMeta:{eventing-e2e9  eventing-e2e9    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Subjects:[]Subject{Subject{Kind:ServiceAccount,APIGroup:,Name:eventing-e2e9,Namespace:eventing-e2e9,},},RoleRef:RoleRef{APIGroup:rbac.authorization.k8s.io,Kind:Role,Name:eventing-e2e9,},}
    creation.go:500: Creating service account &ServiceAccount{ObjectMeta:{eventing-e2e9-eventwatcher  eventing-e2e9    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Secrets:[]ObjectReference{},ImagePullSecrets:[]LocalObjectReference{},AutomountServiceAccountToken:nil,}
    creation.go:528: Creating role &Role{ObjectMeta:{eventing-e2e9-eventwatcher      0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Rules:[]PolicyRule{PolicyRule{Verbs:[get list watch],APIGroups:[],Resources:[pods events],ResourceNames:[],NonResourceURLs:[],},},}
    creation.go:547: Creating role binding &RoleBinding{ObjectMeta:{eventing-e2e9-eventwatcher  eventing-e2e9    0 0001-01-01 00:00:00 +0000 UTC <nil> <nil> map[] map[] [] []  []},Subjects:[]Subject{Subject{Kind:ServiceAccount,APIGroup:,Name:eventing-e2e9-eventwatcher,Namespace:eventing-e2e9,},},RoleRef:RoleRef{APIGroup:rbac.authorization.k8s.io,Kind:Role,Name:eventing-e2e9-eventwatcher,},}
    kafka_sink.go:70: {Topic:kafka-sink-eventing-e2e9 NumPartitions:0xc000c9de3c ReplicationFactor:0xc000c9de4e BootstrapServers:[my-cluster-kafka-bootstrap.kafka:9094] ContentMode:0xc000eada10 Auth:0xc00012c800}
    kafka_sink.go:92: 
        	Error Trace:	kafka_sink.go:92
        	Error:      	Expected nil, but got: &errors.StatusError{ErrStatus:v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"Internal error occurred: failed calling webhook \"defaulting.webhook.kafka.eventing.knative.dev\": failed to call webhook: Post \"[https://kafka-webhook-eventing.knative-eventing.svc:443/defaulting?timeout=2s](https://kafka-webhook-eventing.knative-eventing.svc/defaulting?timeout=2s)\": EOF", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc000db6000), Code:500}}
        	Test:       	TestKafkaSinkV1Alpha1AuthSslSaslScram512/0
```

/retest

devguyio commented:
> The main concerns on merging this PR, 1 day before the release, are
> 
> * the PR is big and touches every component, so periodic jobs might reveal regressions but we don't have time to fix them
> * tests are not passing as of today [1]
> * I reviewed it again and I have additional comments
> * CE SQL implementation isn't stable and deeply tested at all, so even experimenting with this new feature might impact other triggers (for example, the validation in the webhook uses the Go implementation, while our data plane uses the Java one and they might contain subtle differences) in terms of performance and reliability
> 
> [1]:
> 
> ```
> --- FAIL: TestNewTriggerFilters (78.38s)
>     --- PASS: TestNewTriggerFilters/Prerequisite (3.19s)
>         --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/enable_new_trigger_filters_experimental_feature (0.04s)
>         --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/install_one_partition_configuration (0.02s)
>         --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/install_broker (0.08s)
>         --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/broker_is_ready (3.03s)
>         --- PASS: TestNewTriggerFilters/Prerequisite/Kafka_broker/Setup/broker_is_addressable (0.01s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger_subscriber (3.48s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Install_trigger (0.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Requirement/Install_matched_event_sender (3.11s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Requirement/Install_unmatched_event_sender (3.15s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (4.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Exact_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger_subscriber (3.10s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Install_trigger (0.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Requirement/Install_matched_event_sender (3.17s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Requirement/Install_unmatched_event_sender (3.09s)
>     --- FAIL: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Prefix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger_subscriber (3.15s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Install_trigger (0.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Requirement/Install_matched_event_sender (3.10s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Requirement/Install_unmatched_event_sender (3.10s)
>     --- FAIL: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.01s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/Suffix_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger_subscriber (3.14s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Install_trigger (0.03s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Setup/Wait_for_trigger_to_become_ready (3.02s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Requirement/Install_matched_event_sender (3.11s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Requirement/Install_unmatched_event_sender (3.14s)
>     --- FAIL: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST]Triggers_with_new_filters_must_deliver_matched_events (0.00s)
>     --- PASS: TestNewTriggerFilters/New_trigger_filters/CloudEvents_SQL_filter/Assert/[Alpha/MUST_NOT]Triggers_with_new_filters_must_not_deliver_unmatched_events (0.00s)
> ```

1. I understand the concern about having it merged before a release, I'd say the dot-release process is here to address this, but I'm fain if folks are anxious a few days before the release to merge something like this. We've a release now and I waited to comment deliberately until  we have a release to make sure that I'm not sending any mixed signals.
2. FWIW, the tests were because eventing-core needed to be upgraded and the failed nightly jobs of eventing been preventing that. So there's no fix that was introduced after that last review round from a tests POV

> CE SQL implementation isn't stable and deeply tested at all, so even experimenting with this new feature might impact other triggers (for example, the validation in the webhook uses the Go implementation, while our data plane uses the Java one and they might contain subtle differences) in terms of performance and reliability

The experimental feature implementation is making sure that when features are disabled, the experimental feature should have very minimal effect, so from the webhook perspective, if the feature is disabled, the cesql parser won't be used and the codepath will not reach the cesql part anywhere. If you see this happening, let me know, cause that should be fixed. But I tried to make sure to be mindful about that exactly for this reason.

devguyio commented:
@pierDipi code review comments were addressed in the following commits:
* [Data-plane: Addressing review comments](https://github.com/knative-sandbox/eventing-kafka-broker/pull/1922/commits/f70c407b916583082abdcc9bac4ca314b885509e)  [f70c407](https://github.com/knative-sandbox/eventing-kafka-broker/pull/1922/commits/f70c407b916583082abdcc9bac4ca314b885509e)
* [Control-plane: Addressing review comments](https://github.com/knative-sandbox/eventing-kafka-broker/pull/1922/commits/0ff0caf385738c765065f1bf63f8f5722b7bfc95) [0ff0caf](https://github.com/knative-sandbox/eventing-kafka-broker/pull/1922/commits/0ff0caf385738c765065f1bf63f8f5722b7bfc95)

devguyio commented:
/test all

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
-
-    for (final var entry : attributes) {
-      if (!entry.getKey().apply(event).equals(entry.getValue())) {
+    logger.debug("{}: Testing event attributes. Event {}", this.getClass().getSimpleName(), event);
```
The one is important here to identify which sub-filters are being tested in something like an ALL filter. e.g.

```
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.AllFilter - Testing event against ALL filters. Event CloudEvent{id='123-42', source=/api/some-source, type='type', datacontenttype='application/cloudevents+json', dataschema=/api/schema, subject='a-subject-42', time=1985-04-12T23:20:50Z, extensions={}}
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.ExactFilter - Testing event attributes. Event CloudEvent{id='123-42', source=/api/some-source, type='type', datacontenttype='application/cloudevents+json', dataschema=/api/schema, subject='a-subject-42', time=1985-04-12T23:20:50Z, extensions={}}
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.ExactFilter - Event attributes matching succeeded
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.ExactFilter - Testing event attributes. Event CloudEvent{id='123-42', source=/api/some-source, type='type', datacontenttype='application/cloudevents+json', dataschema=/api/schema, subject='a-subject-42', time=1985-04-12T23:20:50Z, extensions={}}
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.ExactFilter - Event attributes matching succeeded
12:38:07.983 [main] DEBUG dev.knative.eventing.kafka.broker.dispatcher.impl.filter.subscriptionsapi.AllFilter - Test ALL filters succeeded. Event CloudEvent{id='123-42', source=/api/some-source, type='type', datacontenttype='application/cloudevents+json', dataschema=/api/schema, subject='a-subject-42', time=1985-04-12T23:20:50Z, extensions={}}
```

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/AllFilter.java:
```diff
@@ -13,27 +13,37 @@
       }
     }
+    logger.debug("{}: Test ALL filters succeeded. Event {}", this.getClass().getSimpleName(),
+      cloudEvent);
```
fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -46,11 +49,12 @@
   // specversion -> 1.0
   // f(event) -> event.getSpecVersion().toString() -> 1.0
-  private final List<Entry<Function<CloudEvent, String>, String>> attributes;
+  private final Map<String, Entry<Function<CloudEvent, String>, String>> attributes;
```
Fixed

devguyio commented on control-plane/pkg/reconciler/trigger/trigger.go:
```diff
@@ -53,7 +54,8 @@ type Reconciler struct {
 
-	Env *config.Env
+	Env   *config.Env
+	Flags *feature.Flags
```
Fixed, no need to be a pointer since it's a map

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
+   * @param given  cloud event current attribute value
+   * @return
+   */
+  public boolean match(String given, String wanted) {
```
fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
-    for (final var entry : attributes) {
-      if (!entry.getKey().apply(event).equals(entry.getValue())) {
+    logger.debug("{}: Testing event attributes. Event {}", this.getClass().getSimpleName(), event);
+    for (final var attribute : attributes.keySet()) {
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
       }
     }
-
+    logger.debug("{}: Event attributes matching succeeded", this.getClass().getSimpleName());
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
-
-    for (final var entry : attributes) {
-      if (!entry.getKey().apply(event).equals(entry.getValue())) {
+    logger.debug("{}: Testing event attributes. Event {}", this.getClass().getSimpleName(), event);
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/AttributesFilter.java:
```diff
@@ -86,16 +90,33 @@ public AttributesFilter(final Map<String, String> attributes) {
+      String existingValue = extractorFunc.apply(event);
+      if (!this.match(existingValue, wantedValue)) {
+        logger.debug("{}: Event attributes matching failed. Attribute: {} Want: {} Got: {} Event: {}",
+          this.getClass().getSimpleName(), attribute, wantedValue, existingValue, event);
```
Fixed

devguyio commented on control-plane/pkg/reconciler/trigger/controller.go:
```diff
@@ -85,6 +87,19 @@ func NewController(ctx context.Context, _ configmap.Watcher, configs *config.Env
+				logger.Warn("Features ConfigMap updated but we didn't get expected flags. Skipping updating cached features")
+			}
+			logger.Debug("Features ConfigMap updated. Updating cached features.")
+			reconciler.Flags = &flags
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/AllFilter.java:
```diff
@@ -13,27 +13,37 @@
   @Override
   public boolean test(CloudEvent cloudEvent) {
+    logger.debug("{}: Testing event against ALL filters. Event {}", this.getClass().getSimpleName(),
+      cloudEvent);
```
Replied above

devguyio commented on control-plane/pkg/reconciler/trigger/trigger.go:
```diff
@@ -53,7 +54,8 @@ type Reconciler struct {
 
-	Env *config.Env
+	Env   *config.Env
+	Flags *feature.Flags
```
Fixed

devguyio commented on control-plane/pkg/reconciler/trigger/controller.go:
```diff
@@ -85,6 +87,19 @@ func NewController(ctx context.Context, _ configmap.Watcher, configs *config.Env
+			if !ok {
+				logger.Warn("Features ConfigMap updated but we didn't get expected flags. Skipping updating cached features")
+			}
+			logger.Debug("Features ConfigMap updated. Updating cached features.")
```
Fixed, can only use the name though, otherwise we'd need to introduce the change in pkg

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/AnyFilter.java:
```diff
@@ -29,11 +33,18 @@ public AnyFilter(Set<Filter> filters) {
       }
     }
+    logger.debug("{}: Test failed. All filters failed. Event {}", this.getClass().getSimpleName(),
+      cloudEvent);
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/CeSqlFilter.java:
```diff
@@ -22,23 +22,33 @@
       return (Boolean) this.runtime.cast(value, Type.BOOLEAN);
     } catch (EvaluationException evaluationException) {
+      logger.error("{}: Exception while evaluating CESQL expression. Test failed. Expression {} - Event {} - Exception {}",
+        this.expression, this.getClass().getSimpleName(), cloudEvent, evaluationException);
```
Fixed

devguyio commented on data-plane/dispatcher/src/main/java/dev/knative/eventing/kafka/broker/dispatcher/impl/filter/subscriptionsapi/AnyFilter.java:
```diff
@@ -29,11 +33,18 @@ public AnyFilter(Set<Filter> filters) {
   public boolean test(CloudEvent cloudEvent) {
+    logger.debug("{}: Testing event against ANY filter. Event {}", this.getClass().getSimpleName(),
+      cloudEvent);
+
```
Replied above

pierDipi reviewed (approved):
/lgtm
/approve

Thank you! :heart_eyes: 

Feel free to remove the `hold`.

devguyio commented:
/unhold

devguyio commented:
Seems like the webhook was unavailable 
```
    broker_trigger_sink_test.go:77: 
        	Error Trace:	broker_trigger_sink_test.go:77
        	Error:      	Expected nil, but got: &errors.StatusError{ErrStatus:v1.Status{TypeMeta:v1.TypeMeta{Kind:"", APIVersion:""}, ListMeta:v1.ListMeta{SelfLink:"", ResourceVersion:"", Continue:"", RemainingItemCount:(*int64)(nil)}, Status:"Failure", Message:"Internal error occurred: failed calling webhook \"validation.webhook.kafka.eventing.knative.dev\": failed to call webhook: Post \"https://kafka-webhook-eventing.knative-eventing.svc:443/resource-validation?timeout=2s\": EOF", Reason:"InternalError", Details:(*v1.StatusDetails)(0xc007792c00), Code:500}}
        	Test:       	TestBrokerV1TriggersV1SinkV1Alpha1/5
```

/retest

aliok reviewed (approved):
/lgtm
/approve

devguyio commented:
/cherry-pick release-1.3
